File: algorithms/py_test/RegisterSvr.py
import sys, glob
import os, time

sys.path.append('../../api_server/gen-py')
sys.path.insert(0, glob.glob('thrift-0.9.3/lib/py/build/lib.*')[0])

from thrift import Thrift
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol

from BigRLab import AlgMgrService
from BigRLab.ttypes import *
from BigRLab.constants import *


# RegisterSvr.py algmgrIp algmgrPort localIp localPort libname servicename nthreads
if __name__ == '__main__':
    #  for arg in sys.argv:
        #  print arg

    algmgrIp = sys.argv[1]
    algmgrPort = int(sys.argv[2])

    algSvrInfo = AlgSvrInfo()
    algSvrInfo.addr = sys.argv[3]
    algSvrInfo.port = int(sys.argv[4])
    algSvrInfo.serviceName = sys.argv[5]
    serviceName = sys.argv[6]
    algSvrInfo.maxConcurrency = int(sys.argv[7])

    # register svr
    algmgrTransport = TSocket.TSocket(algmgrIp, algmgrPort)
    algmgrTransport = TTransport.TFramedTransport(algmgrTransport)
    algmgrProtocol = TBinaryProtocol.TBinaryProtocol(algmgrTransport)
    algmgrCli = AlgMgrService.Client(algmgrProtocol)
    algmgrTransport.open()
    algmgrCli.rmSvr(serviceName, algSvrInfo)
    ret = algmgrCli.addSvr(serviceName, algSvrInfo)
    if ret != 0:
        #  print "Register server fail! retval = %d" % ret
        sys.exit(-1)
    #  print "Register server success!"
    sys.exit(0)

File: algorithms/py_test/RegisterSvr.cpp
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <sstream>
#include <fstream>
#include <iterator>
#include <memory>
#include <thread>
#include <boost/make_shared.hpp>
#include <boost/lexical_cast.hpp>
#include <glog/logging.h>
#include "common.hpp"
#include "rpc_module.h"
#include "alg_common.hpp"
#include "register_svr.hpp"


int main(int argc, char **argv)
try {
    using namespace std;

    // for (int i = 1; i < argc; ++i)
        // cout << argv[i] << " ";
    // cout << endl;

    string          algmgrIp = argv[1];
    uint16_t        algmgrPort = boost::lexical_cast<uint16_t>(argv[2]);

    BigRLab::AlgSvrInfo     algSvrInfo;
    algSvrInfo.addr = argv[3];
    algSvrInfo.port = (int16_t)boost::lexical_cast<uint16_t>(argv[4]);
    algSvrInfo.serviceName = argv[5];
    string          serviceName = argv[6];
    algSvrInfo.maxConcurrency = boost::lexical_cast<int>(argv[7]);

    AlgMgrClient    cli(algmgrIp, algmgrPort);
    cli.start(50, 300);
    cli.client()->rmSvr(serviceName, algSvrInfo);
    cli.client()->addSvr(serviceName, algSvrInfo);

    return 0;

} catch (const std::exception &ex) {
    std::cerr << "Exception caught by main: " << ex.what() << std::endl;
    return -1;
}
